src/lua/minifio.lua

```diff
@@ -0,0 +1,113 @@
+local ffi = require('ffi')
+local minifio = require('internal.minifio')
+
+-- Several file manipulation functions without dependencies on
+-- tarantool's built-in modules to use at early initialization
+-- stage.
+
+ffi.cdef([[
+    char *
+    dirname(char *path);
+]])
+
+-- {{{ Functions exposed by fio
+
+-- NB: minifio.cwd() is defined in src/lua/minifio.c.
+
+function minifio.pathjoin(...)
+    local i, path = 1, nil
+
+    local len = select('#', ...)
+    while i <= len do
+        local sp = select(i, ...)
+        if sp == nil then
+            error("fio.pathjoin(): undefined path part "..i, 0)
+        end
+
+        sp = tostring(sp)
+        if sp ~= '' then
+            path = sp
+            break
+        else
+            i = i + 1
+        end
+    end
+
+    if path == nil then
+        return '.'
+    end
+
+    i = i + 1
+    while i <= len do
+        local sp = select(i, ...)
+        if sp == nil then
+            error("fio.pathjoin(): undefined path part "..i, 0)
+        end
+
+        sp = tostring(sp)
+        if sp ~= '' then
+            path = path .. '/' .. sp
+        end
+
+        i = i + 1
+    end
+
+    path = path:gsub('/+', '/')
+    if path ~= '/' then
+        path = path:gsub('/$', '')
+    end
+
+    return path
+end
+
+function minifio.abspath(path)
+    if path == nil then
+        error("Usage: fio.abspath(path)", 0)
+    end
+    path = path
+    local joined_path
+    local path_tab = {}
+    if string.sub(path, 1, 1) == '/' then
+        joined_path = path
+    else
+        joined_path = minifio.pathjoin(minifio.cwd(), path)
+    end
+    for sp in string.gmatch(joined_path, '[^/]+') do
+        if sp == '..' then
+            table.remove(path_tab)
+        elseif sp ~= '.' then
+            table.insert(path_tab, sp)
+        end
+    end
+    return '/' .. table.concat(path_tab, '/')
+end
+
+-- }}} Functions exposed by fio
+
+-- {{{ Functions replaced by fio
+
+-- Similar to fio.dirname, but it doesn't use
+-- cord_ibuf_take()/cord_ibuf_put().
+function minifio.dirname(path)
```

Comment:
There is one more thing I can't figure out: why both functions (`cwd` and `dirname`) can't be implemented via the one mechanism (either Lua C or FFI)?

I'd rather move `dirname` to minifio.c either in scope of this patch.

Reply:
And we'll end with the question: why both `dirname()` functions can't be implemented via the one mechanish? :)

...and you'll ask me to rewrite `fio.dirname()` as well. But since it is in C anyway, we can just use `cord_ibuf_*()` in C and implement it once in minifio. But will it better this way? We should perform several performance tests: jit on/off, amd64/arm64.

We should stop at some point to actually solve current task instead of some other one and my copy-and-replace-two-calls variant looks as a good candidate to stop on it.

I see a small gain about existing `minifio.dirname()` implementation: it is very similar to `fio.dirname()`, so they're easy to compare visually.

Reply:
OK, let's leave it as is for now.

I guess we need to collect all these points somewhere near Application Server TODO list, so everyone can make this little maintenance from time to time.

Resolving this also.

Reply:
> I guess we need to collect all these points somewhere near Application Server TODO list, so everyone can make this little maintenance from time to time.

I like this idea. I will do something toward this direction.

Reply:
Tracked [here](https://www.notion.so/tarantool/fio-s-approaches-C-vs-Lua-allocation-mechanism-f508ed23eea34732aa97237100aa9b05) now.